From 7897096b61e91063e6afdae1b80b48d45dff75c3 Mon Sep 17 00:00:00 2001
From: Erik Schumacher <erik.schumacher@iris-sensing.com>
Date: Thu, 15 Sep 2022 12:49:55 +0200
Subject: [PATCH 1/3] [RDPHOEN-1314] Fix MIPI settings on tc358746

tc358746.c:
After a soft reset the PLL must be configured before any DHY/CSI related
registers are written.
Minor code fixes.

imx8-mipi-csi2-sam.c:
Add debug = 1 to print summary of communication errors

Signed-off-by: Erik Schumacher <erik.schumacher@iris-sensing.com>
---
 drivers/media/i2c/tc358746.c                  | 28 +++++++++----------
 .../staging/media/imx/imx8-mipi-csi2-sam.c    |  2 +-
 2 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/drivers/media/i2c/tc358746.c b/drivers/media/i2c/tc358746.c
index f5c30a67d321..16c4aaebdd40 100644
--- a/drivers/media/i2c/tc358746.c
+++ b/drivers/media/i2c/tc358746.c
@@ -236,7 +236,7 @@ tc358746_dump_csi(struct v4l2_subdev *sd,
 	v4l2_dbg(2, debug, sd, "twakeupcnt %u\n", csi_setting->twakeupcnt);
 	v4l2_dbg(2, debug, sd, "tclk_postcnt %u\n", csi_setting->tclk_postcnt);
 	v4l2_dbg(2, debug, sd, "ths_trailcnt %u\n", csi_setting->ths_trailcnt);
-	v4l2_dbg(2, debug, sd, "csi_hs_lp_hs_ps %u (%u us)\n",
+	v4l2_dbg(2, debug, sd, "csi_hs_lp_hs_ps %u (%u ns)\n",
 		csi_setting->csi_hs_lp_hs_ps,
 		csi_setting->csi_hs_lp_hs_ps / 1000);
 }
@@ -1143,18 +1143,17 @@ static int tc358746_s_power(struct v4l2_subdev *sd, int on)
 	 */
 	tc358746_sreset(sd);
 
-	if (state->fmt_changed) {
-		tc358746_set_buffers(sd);
-		tc358746_set_csi(sd);
-		tc358746_set_csi_color_space(sd);
+	/* always reconfigure the PLL and CSI registers after a sw reset */
+	/* switch to sleep mode as recommend in REF_01 */
+	tc358746_sleep_mode(sd, 1);
+	tc358746_set_pll(sd);
+	tc358746_sleep_mode(sd, 0);
 
-		/* as recommend in REF_01 */
-		tc358746_sleep_mode(sd, 1);
-		tc358746_set_pll(sd);
-		tc358746_sleep_mode(sd, 0);
+	tc358746_set_buffers(sd);
+	tc358746_set_csi(sd);
+	tc358746_set_csi_color_space(sd);
 
-		state->fmt_changed = false;
-	}
+	state->fmt_changed = false;
 
 	tc358746_enable_csi_lanes(sd, on);
 	tc358746_enable_csi_module(sd, on);
@@ -1595,7 +1594,7 @@ static int tc358746_probe_fw(struct tc358746_state *state)
 	 * The PLL input clock is obtained by dividing refclk by pll_prd.
 	 * It must be between 4 MHz and 40 MHz, lower frequency is better.
 	 */
-	pll_prediv = DIV_ROUND_CLOSEST(refclk, 4000000);
+	pll_prediv = DIV_ROUND_DOWN_ULL(refclk, 4000000);
 	if (pll_prediv < 1 || pll_prediv > 16) {
 		v4l2_err(sd, "Invalid pll pre-divider value: %d\n", pll_prediv);
 		return -EINVAL;
@@ -1850,11 +1849,12 @@ static int tc358746_probe(struct i2c_client *client)
 
 	/* Apply default settings */
 	tc358746_sreset(sd);
+	tc358746_sleep_mode(sd, 1);
+	tc358746_set_pll(sd);
+
 	tc358746_set_buffers(sd);
 	tc358746_set_csi(sd);
 	tc358746_set_csi_color_space(sd);
-	tc358746_sleep_mode(sd, 1);
-	tc358746_set_pll(sd);
 	tc358746_enable_stream(sd, 0);
 
 	err = tc358746_async_register(state);
diff --git a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
index 43656a1daa02..824af1321370 100644
--- a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
+++ b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
@@ -406,7 +406,7 @@ struct csi_state {
 	bool hdr;
 };
 
-static int debug;
+static int debug = 1;
 module_param(debug, int, 0644);
 MODULE_PARM_DESC(debug, "Debug level (0-2)");
 
-- 
2.37.3

